docs/node-mixin/rules/rules.libsonnet

```diff
@@ -53,9 +53,9 @@
             record: 'instance:node_memory_swap_io_pages:rate1m',
             expr: |||
               (
-                rate(node_vmstat_pgpgin{%(nodeExporterSelector)s}[1m])
+                rate(node_vmstat_pswpin{%(nodeExporterSelector)s}[1m])
               +
-                rate(node_vmstat_pgpgout{%(nodeExporterSelector)s}[1m])
+                rate(node_vmstat_pswpout{%(nodeExporterSelector)s}[1m])
               )
             ||| % $._config,
```

Comment:
Thanks @SuperQ for the input. `node_vmstat_pgmajfault` seems to me like the ideal metric to represent what we want here, namely some notion of memory saturation. (Major page faults will also happen in the absence of any memory pressure if a file is mmap'd, which is ironically done heavily by Prometheus itself. That's probably the reason why the Gitlab alert also required a lack of available memory. But this is not an alert, just for a dashboard graph.)

My suggestion for the rule would be:

```jsonnet
           record: 'instance:node_vmstat_pgmajfault:rate1m',
           expr: |||
             rate(node_vmstat_pgmajfault{%(nodeExporterSelector)s}[1m])
           ||| % $._config,
```

The dashboard needed to be adjusted accordingly because the name of the rule has changed.

What does everyone think?

Reply:
I thought that Prometheus would only cause normal minor page faults, not major faults for the mmap data.

Reply:
IIUC a minor page fault happens when memory is allocated but not yet mapped to physical RAM. If that memory is accessed, the kernel counts that as a minor page fault and assign a page in physical RAM. A major page fault happens when accessing a page that is physically on disk, which happens if a page was written to disk because of memory pressure, or a file was mmap'd where the content was not in RAM before anyway. (In my understanding, if Prometheus writes a new block, all the data is in the page cache anyway because the file was just written to disk. However, if Prometheus accesses an older block, or it comes up again after a machine reboot, all the block reads should initially trigger major page faults.)
But I might be totally wrong. That's just what I got from reading stackoverflow etc.